src/include/miopen/gather/problem_description.hpp

```diff
@@ -0,0 +1,166 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#pragma once
+
+#include <cstddef>
+#include <cstdint>
+
+#include <miopen/errors.hpp>
+#include "miopen/gather.hpp"
+#include <miopen/tensor.hpp>
+#include <miopen/problem_description_base.hpp>
+#include "../src/kernels/tensor_view.hpp"
+
+namespace miopen {
+
+struct NetworkConfig;
+
+namespace gather {
+
+template <int M>
+tensor_view_t<M> reshape(const TensorDescriptor& tensorDes, const std::vector<size_t>& shape);
+
+extern template tensor_view_t<3> reshape<3>(const TensorDescriptor& tensorDes,
+                                            const std::vector<size_t>& shape);
+
+extern template tensor_view_t<4> reshape<4>(const TensorDescriptor& tensorDes,
+                                            const std::vector<size_t>& shape);
+
+struct BwdProblemDescription : ProblemDescriptionBase
+{
+    // Backward constructor
+    BwdProblemDescription(const GatherDescriptor& gatherDesc_,
+                          const TensorDescriptor& outputGradDesc_,
+                          const TensorDescriptor& indicesDesc_,
+                          const TensorDescriptor& paramGradDesc_)
+        : gatherDesc(gatherDesc_),
+          outputGradDesc(outputGradDesc_),
+          indicesDesc(indicesDesc_),
+          paramGradDesc(paramGradDesc_)
+    {
+        auto dim        = gatherDesc.getDim();
+        auto batch_dims = gatherDesc.getBatchDims();
+
+        if(indicesDesc.GetType() != miopenInt32 && indicesDesc.GetType() != miopenInt64)
+        {
+            MIOPEN_THROW(
+                miopenStatusBadParm,
+                "GatherV2::BwdProblemDescription: Indices tensor must be int32 or int64 tensor.");
+        }
+
+        if(!IsSameType())
+        {
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "GatherV2::BwdProblemDescription: Output grad and param grad tensors must "
+                         "be same type.");
+        }
+
+        if(!IsAllContiguous())
+        {
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "GatherV2::BwdProblemDescription: All tensors must be contiguous.");
+        }
```

Comment:
Does this mean you only implemented contiguous version of gather?

Reply:
Yes, only contiguous version is provided in modnn

Reply:
So this should be in IsApplicable function, not the problem of the op

Reply:
and add IsAllContiguous to [src/gather/problem_description.cpp](https://github.com/moreh-dev/MIOpen/pull/55/files#diff-74de9c76955ab72545c4ff7c674f7d3fbbaa2d31a8af94481759fd0df1329e2c) also

Reply:
Actually, this check can remain either within the problem's description or the `IsApplication` function without any impact. I placed it within the problem's description to avoid needing to add the contiguous check to the kernel cache

Reply:
I think this is not a BadParm, just because your solver dont support noncontiguous, it should be return false, not MIOPEN_THROW and be in IsApplication to make it clear meaning

Reply:
I changed error code to `NotImplemented`. `MIOPEN_THROW` or `return false` basically stop the function from executing but I think keeping `MIOPEN_THROW` would be more align to existing coding conventions. I also give proper message in `MIOPEN_THROW` for clear meaning.